### PR TITLE
Support for wildcard resources

### PR DESCRIPTION
**Pull request by chrisdunelm**

### CODE REVIEW

File: Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNamesClient.cs [hunk 1]
```diff
@@ -254,6 +254,155 @@ public abstract class ResourceNamesClient
...
+        public virtual Response WildcardMethod(string name, scg::IEnumerable<string> names, string requiredSingle, scg::IEnumerable<string> requiredRepeated, gaxgrpc::CallSettings callSettings = null) =>
+            WildcardMethod(new WildcardResource
+            {
+                Name = name ?? "",
+                Names =
+                {
+                    names ?? linq::Enumerable.Empty<string>(),
+                },
+                RequiredSingle = gax::GaxPreconditions.CheckNotNullOrEmpty(requiredSingle, nameof(requiredSingle)),
+                RequiredRepeated =
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNamesClient.cs]: I'm always somewhat dubious of "required repeated" - it's not like we're checking that it's non-empty. Probably worth further consideration.

File: Google.Api.Generator/Generation/ResourceNamesGenerator.cs [hunk 2]
```diff
@@ -235,15 +237,14 @@ private IEnumerable<ClassDeclarationSyntax> ProtoMessagePartials()
                                 if (res.field.IsRepeated)
                                 {
                                     var repeatedTyp = Typ.Generic(typeof(ResourceNameList<>), one.ResourceNameTyp);
+                                    var s = Parameter(null, "s");
```
Comment (jskeet) [inline Google.Api.Generator/Generation/ResourceNamesGenerator.cs]: Move this into the IsWildcard scope? I don't think it's needed outside it.

File: Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNamesClient.cs (hunk 1 above)
Comment (chrisdunelm) [reply Google.Api.Generator.Tests/ProtoTests/ResourceNames/ResourceNamesClient.cs]: Yes, I agree, and we may want to discuss changing how we support this.
However, this matches the existing code-gen, so is "correct".

File: Google.Api.Generator/Generation/ResourceNamesGenerator.cs (hunk 2 above)
Comment (chrisdunelm) [reply Google.Api.Generator/Generation/ResourceNamesGenerator.cs]: `s` is also used in the `else` clause. The unhelpfully short name doesn't help.

**Pull request merged**